evmrpc/filter.go

```diff
@@ -805,6 +834,47 @@ func (f *LogFetcher) mergeSortedLogs(batches [][]*ethtypes.Log) []*ethtypes.Log
 	return res
 }
 
+func (f *LogFetcher) ensureHeightAvailable(ctx context.Context, height int64) error {
+	if f.watermarks == nil {
+		return nil
+	}
+	return f.watermarks.EnsureHeightAvailable(ctx, height)
+}
+
+func (f *LogFetcher) latestHeight(ctx context.Context) (int64, error) {
+	if f.watermarks != nil {
+		return f.watermarks.LatestHeight(ctx)
+	}
+	if f.ctxProvider == nil {
+		return 0, fmt.Errorf("ctx provider not configured")
+	}
+	return f.ctxProvider(LatestCtxHeight).BlockHeight(), nil
+}
+
+func (f *LogFetcher) earliestHeight(ctx context.Context) (int64, error) {
+	if f.watermarks != nil {
+		return f.watermarks.EarliestHeight(ctx)
```

Comment:
Can we always assume and use watermark to get earliest version? Thus avoid all the below logic?